docs: add how-to-write-a-module.md (#456)
netdata/go.d.plugin: docs/how-to-write-a-module.md

| [`module_name.go`](#file-module_namego)           | Module configuration, implementation and registration. |
| [`charts.go`](#file-chartsgo)                     | Charts, charts templates and constructor functions.    |
| [`init.go`](#file-initgo)                         | Initialization methods.                                |
| [`collect.go`](#file-collectgo)                   | Metrics collection implementation.                     |
| [`module_name_test.go`](#file-module_name_testgo) | Public methods/functions tests.                        |
| [`testdata/`](#file-module_name_testgo)           | Tests fixtures.                                        |

### File `module_name.go`

> :exclamation: See the example [`examlpe.go`](https://github.com/netdata/go.d.plugin/blob/master/modules/example/example.go).

Don't overload this file with the implementation details.

Usually it contains only:
 
-   module registration.
-   module configuration.
-   [module interface implementation](#module-interface).

### File `charts.go`

> :exclamation: See the example: [`charts.go`](https://github.com/netdata/go.d.plugin/blob/master/modules/example/charts.go).

> Put charts, charts templates and charts constructor functions in this file. 

Author: > See the example: [`charts.go`]

> Put charts, charts templates and charts constructor functions in this file. 

Reviewer: I think the document should be all-encompassing, meaning that I don't need to go here and there to get required information to complete my understanding. Furthermore, I still can't understand what those data structure are or what is their purpose. 
 Let's get as explicit as possible, ELI5 level.

> Put charts, charts templates and charts constructor functions in this file. 

Reviewer: Our goal is to lower the barrier of entry as much as possible, so that people can contribute easier and with less required technical understanding.

> Put charts, charts templates and charts constructor functions in this file. 

Author: We expect people checking other modules source code and source code in general (it is very easy to do useing IDE these days), it is impossobile ot avoid it.

You think that this is difficult, but it is not.

> Put charts, charts templates and charts constructor functions in this file. 

Reviewer: I checked code you mean, but what I can't understand is what those data structures means for module, from a logical perspective.

The idea is to allow even rookie go programmers to easily grasp the fundamentals and create modules, otherwise, people will continue using python.

> Put charts, charts templates and charts constructor functions in this file. 

Reviewer: I checked the file. I want us to explain to users what those mean in the netdata context, what is a chart and what is a template. Why in the example, one is hidden, what's the functionality of a hidden chart (i.e why would I want to do it). 
Give me all the information so that I don't need to make **any** assumption. Allow me to focus on business logic and now to understand the netdata aspects.

# How to write a module

> Let's assume you want to write `example` module.

Reviewer 2: ```suggestion
Let's assume you want to write a collector named `example`.
```

Steps are:

> - add the source code to the [`modules/example/`](https://github.com/netdata/go.d.plugin/tree/master/modules).
>   - [module interface](#module-interface).
>   - [suggested module layout](#module-layout).
>   - [helper packages](#helper-packages).
> - add the configuration to the [`config/go.d/exmaple.conf`](https://github.com/netdata/go.d.plugin/tree/master/config/go.d).
> - add the module to the [`config/go.d.conf`](https://github.com/netdata/go.d.plugin/blob/master/config/go.d.conf).
> - import the module in [`modules/init.go`](https://github.com/netdata/go.d.plugin/blob/master/modules/init.go).
> - update the [`available modules list`](https://github.com/netdata/go.d.plugin#available-modules).

Reviewer 2: ```suggestion
- Add the source code to [`modules/example/`](https://github.com/netdata/go.d.plugin/tree/master/modules).
  - [module interface](#module-interface).
  - [suggested module layout](#module-layout).
  - [helper packages](#helper-packages).
- Add the configuration to [`config/go.d/example.conf`](https://github.com/netdata/go.d.plugin/tree/master/config/go.d).
- Add the module to [`config/go.d.conf`](https://github.com/netdata/go.d.plugin/blob/master/config/go.d.conf).
- Import the module in [`modules/init.go`](https://github.com/netdata/go.d.plugin/blob/master/modules/init.go).
- Update the [`available modules list`](https://github.com/netdata/go.d.plugin#available-modules).
```

-   Called after `Check` and only if `Check` returned `true`.
-   If it returns `nil`, the job will be disabled
-   :warning: Make sure not to share returned value between module instances (jobs).

Usually charts initialized in `Init` and `Chart` method just returns the charts instance:

```go
// example.go

func (e *Example) Charts() *Charts {
    return e.charts
}
```

### Collect method

-   `Collect` collects metrics.
-    Called only if `Check` returned `true`.
-    Called every `update_every` seconds.
-    `map[string]int64` keys are charts dimensions ids'.

We propose to use the following template:
 
```go
// example.go

func (e *Example) Collect() map[string]int64 {
    ms, err := e.collect()
    if err != nil {
        e.Error(err)
    }

    if len(ms) == 0 {
        return nil
    }
    return ms
}
```

Move metrics collection logic into the `collect.go` file. See [suggested module layout](#module-Layout).

### Cleanup method

-   `Cleanup` performs the job cleanup/teardown.
-    Called if `Init` or `Check` fails, or we want to stop the job after `Collect`.

If you have nothing to clean up:

```go
// example.go

func (Example) Cleanup() {}
```

## Module Layout

> The general idea - do not put everything in a file.

Reviewer 2: ```suggestion
The general idea is to not put everything in a single file.
```

> We recommend using a file per a logical area. This approach makes it easier to maintain the module.

Reviewer 2: ```suggestion
We recommend using one file per logical area. This approach makes it easier to maintain the module.
```

Suggested minimal layout:

| Filename                                          | Contains                                                |
| ------------------------------------------------- |-------------------------------------------------------- |
| [`module_name.go`](#file-module_namego)           | Module configuration, implementation and registration.  |
| [`charts.go`](#file-chartsgo)                     | Charts, charts templates and constructor functions.     |
| [`init.go`](#file-initgo)                         | Initialization methods.                                 |
| [`collect.go`](#file-collectgo)                   | Metrics collection implementation.                      |
| [`module_name_test.go`](#file-module_name_testgo) | Public methods/functions tests.                         |
| [`testdata/`](#file-module_name_testgo)           | Files containing sample data.                           |

### File `module_name.go`

> > :exclamation: See the example [`examlpe.go`](https://github.com/netdata/go.d.plugin/blob/master/modules/example/example.go).

Reviewer: ```suggestion
> :exclamation: See the example [`example.go`](https://github.com/netdata/go.d.plugin/blob/master/modules/example/example.go).
```

All the module initialization details should go in this file.

-   make a function for each value that needs to be initialized.
-   a function should return a value(s), not implicitly set/change any values in the main struct.

```go
// init.go

// Prefer this approach.
func (e Example) initSomeValue() (someValue, error) {
    // ...
    return someValue, nil 
}

// This approach is ok too, but we recommend to not use it.
func (e *Example) initSomeValue() error {
    // ...
    m.someValue = someValue
    return nil
}
```     

### File `collect.go`

> :exclamation: See the example: [`collect.go`](https://github.com/netdata/go.d.plugin/blob/master/modules/example/collect.go).

This file is the entry point for the metrics collection.

Feel free to split it into several files if you think it makes the code more readable.

Use `collect_` prefix for the filenames: `collect_this.go`, `collect_that.go`, etc.

```go
// collect.go

func (e *Example) collect() (map[string]int64, error) {
    collected := make(map[string])int64
    // ...
    // ...
    // ...
    return collected, nil
}
```

> ### File `module_name_test.go`

Reviewer: Can we add some links to resources for creating tests in go?  Can we give some examples from already implemented collectors? 
Examples would be test code + explanation/description of what it does. I want even the go rookie programmer who hasn't written any tests to quickly understand the concept/logic and create the tests without giving up on the collector.
Remember, we want to nudge people from creating a quick and dirty python collector to using go which is more complex but more robust as well. 
Perhaps @kaskavel can help with this one?

Reviewer 3: @OdysLam I will take a look now on this. Let me go through the existing documentation and I 'll post here.

> :exclamation: See the example: [`example_test.go`](https://github.com/netdata/go.d.plugin/blob/master/modules/example/example_test.go).

Testing is mandatory.

-   test only public functions and methods (`New`, `Init`, `Check`, `Charts`, `Cleanup`, `Collect`).
-   do not create a test function per a case, use [table driven tests](https://github.com/golang/go/wiki/TableDrivenTests).
Prefer `map[string]struct{ ... }` over `[]struct{ ... }`.
-   use helper functions _to prepare_ test cases to keep them clean and readable.  

### Directory `testdata/`

> The directory contains tests fixtures.

Reviewer: What are tests fixtures? Let's assume no knowledge from the part of the user.
